cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -1700,14 +1802,30 @@ struct GroupedMinMaxImpl : public GroupedAggregator {
     auto raw_mins = reinterpret_cast<CType*>(mins_.mutable_data());
     auto raw_maxes = reinterpret_cast<CType*>(maxes_.mutable_data());
 
-    VisitArrayValuesInline<Type>(
-        *batch[0].array(),
-        [&](CType val) {
+    if (batch[0].is_array()) {
+      VisitArrayValuesInline<Type>(
+          *batch[0].array(),
+          [&](CType val) {
+            raw_maxes[*g] = std::max(raw_maxes[*g], val);
+            raw_mins[*g] = std::min(raw_mins[*g], val);
+            BitUtil::SetBit(has_values_.mutable_data(), *g++);
+          },
+          [&] { BitUtil::SetBit(has_nulls_.mutable_data(), *g++); });
+    } else {
+      const auto& input = *batch[0].scalar();
+      if (input.is_valid) {
+        const auto val = UnboxScalar<Type>::Unbox(input);
+        for (int64_t i = 0; i < batch.length; i++) {
           raw_maxes[*g] = std::max(raw_maxes[*g], val);
           raw_mins[*g] = std::min(raw_mins[*g], val);
           BitUtil::SetBit(has_values_.mutable_data(), *g++);
-        },
-        [&] { BitUtil::SetBit(has_nulls_.mutable_data(), *g++); });
+        }
+      } else {
+        for (int64_t i = 0; i < batch.length; i++) {
+          BitUtil::SetBit(has_nulls_.mutable_data(), *g++);
+        }
+      }
+    }
```

Comment:
So it seems it would be nicer to write this as:
```c++
  auto consume_value = [&](uint32_t g, CType val) {
    raw_maxes[g] = std::max(raw_maxes[g], val);
    raw_mins[g] = std::min(raw_mins[g], val);
    BitUtil::SetBit(has_values_.mutable_data(), g);
  };
  auto consume_null = [&](uint32_t g) {
    BitUtil::SetBit(has_nulls_.mutable_data(), g);
  };
  VisitGroupedValues<Type>(batch, std::move(consume_value), std::move(consume_nulls));
```

Reply:
You gave me hope that a helper like this already existed :( But that is a good idea, it would clean things up quite a bit.